rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,272 @@
+
+##### PIN Prompt
+
+Once the key is generated, we will need to prompt the user to enter a new PIN for it. If the PIN policy is `never`, then this prompt will be skipped.
+
+TBD: need to flush out how this will interact with webauthn/etc.
+
+##### Management Key
+
+TBD
+
+### yubikey login session
+
+Since `tsh` cannot detect a private key on disk like usual, we'll need to provide a way for `tsh` to reaccess its login session in another way.
+
+On yubikey login, `tsh` will add `yubikey: true` to the user's `~/.tsh/proxy.yaml` file so that `tsh` knows to use the user's yubikey for subsequent `tsh` commands. If the yubikey or its private key is not found or does not match the certificates on file, the user will be receive an error or be prompted to re-login like usual.
+
+### Server Enforcement
+
+In order to make this functionality from the server's perspective, we need to add ways to enforce yubikey login and guarentee its proper usage. This can be done with [Attestation](https://docs.yubico.com/yesdk/users-manual/application-piv/attestation.html), which we can use to verify the certificate chain from the yubikey's public key to the yubikey's CA.
+
+For this [verification](https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#Verify) to take place on the server, we will need to pass the user's yubikey [attestation certificate](https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#YubiKey.AttestationCertificate), as well as an attestation statement [slot cert](https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#YubiKey.Attest) for the slot in question. Once this verification succeeds, we can be positive that the public key in the slot cert belongs to a yubikey-backed private key. If this public key matches the one the Auth Server is preparing to sign, then it can safely continue to do so.
+
+To pass the attestation certificates to the Auth Server, they will be added to the login endpoints, including `/webali/ssh/certs`, `/webapi/mfa/login/finish`, `/webapi/<sso>/callback`.
+
+Additionally, the attestation will return useful information about the user's yubikey and the private key in question, including:
+ - The private key's PIN policy
+ - The private key's Touch policy
+ - The yubikey serial number
+
+#### Auth Service Config or Auth Preference
+
+To require yubikey login for all users, you can set the `require_yubikey_login` field in the Cluster's Auth Preference, either through the Auth Service's config file:
+
+```yaml
+auth_service:
+  ...
+  authentication:
+    ...
+    yubikey_login: 
+      required: true
+      pin_policy: always
+      touch_policy: always
+```
+
+Or with a dynamic Cluster Auth Preference object:
+
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  yubikey_login: 
+    required: true
+    pin_policy: always
+    touch_policy: always
+```
+
+If `yubikey_login_required` is `false`, the policy fields will be ignored and attestation will be skipped.
```

Comment:
My 2c: if you can you should still get the attestation statement, verify it and save it in storage. That allows us to make assertions about the keys after the fact, which could come in handy for config changes or new Teleport versions.